File: src/spa.rs
use std::fs;
use std::path::Path;
use std::sync::Mutex;

static CACHED_HTML: Mutex<Option<(String, u64)>> = Mutex::new(None);

pub fn get_index_html(base_path: &str) -> Option<String> {
    let index_path = Path::new("client/build/index.html");

    // In production, return cached version
    let is_prod = std::env::var("NODE_ENV")
        .map(|v| v == "production")
        .unwrap_or(false);

    let mut cached = CACHED_HTML.lock().unwrap();

    let metadata = fs::metadata(index_path).ok()?;
    let mtime = metadata
        .modified()
        .ok()?
        .duration_since(std::time::UNIX_EPOCH)
        .ok()?
        .as_millis() as u64;

    if let Some((ref html, ref cached_mtime)) = *cached {
        if is_prod || mtime == *cached_mtime {
            return Some(html.clone());
        }
    }

    let raw = fs::read_to_string(index_path).ok()?;
    let html = raw
        .replace("<head>", &format!("<head>\n\t\t<base href=\"{}/\">", base_path))
        .replace(
            "window.__BASE_PATH__ = \"\"",
            &format!(
                "window.__BASE_PATH__ = {}",
                serde_json::to_string(base_path).unwrap()
            ),
        );

    *cached = Some((html.clone(), mtime));
    Some(html)
}

File: src/spa.nim
import std/[os, json, strutils, times, options]

var
  cachedHtml: string = ""
  cachedMtime: Time = Time()

proc getIndexHtml*(basePath: string): Option[string] =
  let indexPath = "client/build/index.html"

  if not fileExists(indexPath):
    return none(string)

  let isProd = getEnv("NODE_ENV") == "production"
  let info = getFileInfo(indexPath)
  let mtime = info.lastWriteTime

  if cachedHtml.len > 0:
    if isProd or mtime == cachedMtime:
      return some(cachedHtml)

  let raw = readFile(indexPath)
  var html = raw.replace("<head>",
    "<head>\n\t\t<base href=\"" & basePath & "/\">")
  html = html.replace(
    "window.__BASE_PATH__ = \"\"",
    "window.__BASE_PATH__ = " & $(%basePath))

  cachedHtml = html
  cachedMtime = mtime
  some(html)
